Analytical sedov

jaescartin1 opened the pull request:
Added analytical solution for the Sedov test.

It generate a file with the solution in 1D for the next geometries: planar(1D), cylindrical(2D) or spherical(3D)

jaescartin1 committed:
Fixed warning: unused variable

jaescartin1 committed:
Add Sedov Solution:
- First attept. Solution conversion from fortran to c++

jaescartin1 committed:
Removed warnings.
Next step remove linking errors

jaescartin1 committed:
Add predeclaration for linking

jaescartin1 committed:
Cosmetics

jaescartin1 committed:
Add test with the original 2D fortran parameters.
Only for compare the original code with the new code

jaescartin1 committed:
Add the fortran code to compare

jaescartin1 committed:
Change time in the test

jaescartin1 committed:
Fixed qromo function

jaescartin1 committed:
Set exponential format in the output

jaescartin1 committed:
Improve the output format

jaescartin1 committed:
Searching diferences

jaescartin1 committed:
Check

jaescartin1 committed:
It seems that is fixed the solution in 2D

jaescartin1 committed:
Added normalized outputs

jaescartin1 committed:
Added normalized density output

jaescartin1 committed:
Commented the 2D test and enable the normal execution

jaescartin1 committed:
Removed analytical solution in Fortran

jaescartin1 committed:
Remove files from the index

jaescartin1 committed:
Remove unused variables

jaescartin1 committed:
Added simulation output, in order to compare with the solution

jaescartin1 committed:
.

jaescartin1 committed:
Fix warning

jaescartin1 committed:
Move write file to the data layer in SedovDataFileWriter.hpp

jaescartin1 committed:
Add analytical solution in Fortran, the original Fortran code

sekelle commented:
Hi Jose, thanks for your work. I'll be excited to see the comparison with an analytical reference.
I think the analytical solution should be an independent app, not coupled to `sedov.cpp` . You could then have a python or bash script that combines the data from the sedov app and the analytical reference and compare them.

jaescartin1 commented:
Hi Sebastian, the comparative is not completely ready yet. The solution works with a planar (1D), cylindrical (2D) or spherical (3D) symmetry.  It means that we still need to modify the sedov sph-exa input to provide a spherical symmetry 3D model. This is the previous step to work with Ruben an reach the final solution.

jaescartin1 commented:
I attached the original fortran code , fully tested in the folder 'src/sedov/analytical', we should remove this folder in the future.
The 'SedovAnalyticalSolution.hpp' file is my c++ adaptation from the original code.

jaescartin1 commented:
Ruben asked me to integrate the Sedov solution inside the sph-exa code, in order to generate automatically the solution when the user select dump and provide the solution flag. It is much more easy and useful to compare the simulate and the analytical output because the solution use automatically the same input parameters than sph-exa on that iteration and time.

jaescartin1 commented:
If you think that everything seems correct, and you don't mind, please approve the pull request and we can modify the implementation in the future if it is need. I would like to have it in the develop branch for the moment to test it the code with Ruben.

jaescartin1 committed:
Added code to compare exactly the output of the original Fortran code
and the C++ adapt code in the Sedov analytical solution

jaescartin1 committed:
Storage internal energy (u) as log10(u) in sim and sol

jaescartin1 committed:
Fix precision in the analytical solution. 1000 steps is enough to
compare

jaescartin1 committed:
Set by default the sedov analytical solution test to 3D and rho=10000

jaescartin1 committed:
Update

jaescartin1 committed:
Add spheric model

jaescartin1 committed:
Duplicate the solution distance in spheric geometry too

jaescartin1 committed:
Changed to cubic model

jaescartin1 committed:
clean include

jaescartin1 committed:
Changed to rMax(r1) dependency in the particle distribution

jaescartin1 committed:
Check spheric model symmetry

jaescartin1 committed:
Reduce the solution to rMax=0.5, that match with the model.

Increased the default iterations from 10 to 100

jaescartin1 committed:
Add paper. It has to be removed in the future

jaescartin1 committed:
Fix Sedov solution. In the solution 'eblast' was 'energytot', not
'ener0'

jaescartin1 committed:
update the default iterations from 100 to 300

jaescartin1 committed:
Merge branch 'develop' into analytical_sedov

jaescartin1 committed:
remove a linebreak

jaescartin1 committed:
Added a couple of [[maybe_unused]] to avoid warnings

jaescartin1 commented:
Hi Sebastian, the Sedov exact solution is in place and working. We can merge the branch.

jgphpc committed:
launching ci

jgphpc committed:
launch ci

jgphpc commented:
cscs-ci run

jgphpc commented:
Some tests failed the CI: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/SPH-EXA-44692846847247/-/pipelines/452889762

jaescartin1 committed:
Remove function: reduction_sphere .
The particle reduccion didn't match well in the GPU code, after the new
setup.

jaescartin1 commented:
Hi Jean-Guillaume,
      thanks for check it.  I just removed a function to reduce the size to a sphere. It seems didn't match well with the GPU implementation after the reduccion. Like this function is not needed for the test, I just removed it.

I hope now the modifications pass the tests.

jgphpc commented:
cscs-ci run

sekelle reviewed (changes requested):
Hi @jaescartin1, thanks for your contribution.
The analytical solution itself and the changes to the SedovDataGenerator are fine.

One problem that remains is that the code for the analytical solution is introduced into the main compilation unit of sedov.cpp.
We have to avoid this coupling, because what's now just the sedov test case in sedov.cpp will evolve into a more general simulation driver that will be able to handle any hydro-only type of simulation, including the turbulence tests on LUMI.
We don't actually want to repeat the whole time-stepping loop for every new set of initial conditions. Therefore we can't have
code for the analytical solution introduced in sedov.cpp

The sedov analytical solution can and should be a stand-alone application: You first execute a run with the sedov application, writing snapshots at the times where you want to compare with the analytical solutions.
Next, sedov-analytical can be executed. It includes the same sedov_data_generator.hpp file as sedov.cpp to get the simulation parameters and it reads the required simulation data (x,y,z,vx,vy,vz,time) from the output that the sedov solution has generated.

sekelle commented:
Note, I would set up `sedov-analytical` such that it compares only a single time/snapshot and control comparing multiple times with a python script.

sekelle commented on src/sedov/SedovDataGenerator.hpp:
```diff
@@ -12,6 +12,16 @@ template <typename T, typename I>
+    static constexpr double energytot     = 1.0;
+    static constexpr double width         = 0.10;
+    static constexpr double rho0          = 1.0;
+    static constexpr double ener0         = energytot / std::pow(M_PI,(3.0/2.0)) / 1.0 / std::pow(width,3);
```
How is this not generating a compiler error in the CI? This should not compile because `std::pow` is not constexpr,
and in fact does not compile for me.

jaescartin1 replied:
I couldn't see this problem with make or cmake in my Ubuntu.
Do you use some additional flags for the compilation?

Maybe is a different compiler version. This is mine:
$ mpic++ --version
g++ (Ubuntu 11.2.0-7ubuntu2) 11.2.0

jaescartin1 replied:
As far as I can see on my system, the pow(x,y) function is taken from the include <math.h> at '/usr/x86_64-linux-gnu/include/bits/mathcalls.h'

$ cat /usr/x86_64-linux-gnu/include/bits/mathcalls.h
...
/* Return X to the Y power.  */
__MATHCALL_VEC (pow,, (_Mdouble_ __x, _Mdouble_ __y));
...

sekelle replied:
It works with GCC and fails with clang. According to the C++17 standard, std::pow is not constexpr, so GCC is wrong and clang is right. This should be:
`static inline const double energ0 = ...`

sekelle replied:
@jgphpc are we only testing with GCC?

jgphpc replied:
yes but I can add clang

jaescartin1 committed:
Changed 'constexpr' by 'inline const'

jaescartin1 committed:
Starting the independence of the analyticalSolution

jaescartin1 committed:
Working in the split of the analytical sedov solution

jaescartin1 committed:
Fixed the external sol sedov exection

jaescartin1 committed:
Added cmake support to the sedovSolution

jaescartin1 committed:
Added OutPath support

jaescartin1 committed:
Add initial python script for compare analytical solutions

jgphpc commented:
cscs-ci run

jaescartin1 committed:
Improving python script

jaescartin1 committed:
Added graphics. It seems working with matplotlib

jaescartin1 committed:
Clean and add some info

jaescartin1 committed:
Cosmetics

jaescartin1 commented:
The required changes are in-place. Please, check it again.

jgphpc commented:
cscs-ci run

sekelle commented on src/analyticalSolutions/compare_solutions.py:
```diff
@@ -0,0 +1,304 @@
+        plt.figure().clear()
+        print("'Radius vs Density' done.")
+    
+        figureName = "./sedov_presure_comparation_"   + time.__str__() + ".svg"
```
perhaps `sedov_pressure` is enough (note double ss) ?

sekelle commented:
Some observations:

- An SVG file with 125000 (and more) particles is computationally very demanding to render, we should perhaps make .png the default
- File naming convention: the most widely used naming scheme is `lower_case_with_underscore`, e.g. `sedov_solution.cpp`, I already applied this convention everywhere in `domain/`, but not yet in `include/`
- I invoked the python script with `./compare_solutions.py sedov --binary_file ./sedovSolution --nparts 125000 --snapshot_file ./dump_sedov200.txt --constants_file ./constants.txt`, i.e. I forgot to also supply `--iteration 200`. In this case the script assumed iteration 0 as the default, so I got a wrong plot. Perhaps we should make `iteration` a mandatory argument if `constants_file` is given

sekelle commented:
When building with cmake, the `compare_solutions.py` script is not copied to the installation directory, so I had to copy that by hand. Perhaps @jgphpc could advise how to modify CMakeLists.txt to also install the python script?

jgphpc commented on src/analyticalSolutions/compare_solutions.py:
```diff
@@ -0,0 +1,304 @@
+        
+        print("\nGenerating graphics ...")
+        
+        figureName = "./sedov_density_comparation_"   + time.__str__() + ".svg"
```
comparison

sekelle replied:
also here, perhaps we can just use `sedov_density_` for brevity.

jgphpc commented on src/analyticalSolutions/sedovSolution/main.cpp:
```diff
@@ -0,0 +1,181 @@
+ * @brief This program generate the analytical sedov solution based in the time and the initial conditions
+ *
+ * @author Jose A. Escartin <ja.escartin@gmail.com>
+ *
```
You may want to add a reference:
https://cococubed.com/papers/kamm_2000.pdf

jaescartin1 replied:
Hi, yes sure, but this is already added but not in the 'main.ccp' where only is called. The reference is in the solution header itself 'SedovSolution.hpp'

 *        This routine is a C++ conversion of one Fortran code based in these two papers:
 *        - "evaluation of the sedov-von neumann-taylor blast wave solution", Jim Kamm, la-ur-00-6055
 *        - "the sedov self-similiar point blast solutions in nonuniform media", David Book, shock waves, 4, 1, 1994

jaescartin1 replied:
The Frank Timmes code (in the source that you pointed out) is a evolution, wrote in 2019, of the base code that I used.
https://cococubed.com/research_pages/sedov.shtml

jaescartin1 replied:
I ported and removed the bad practices (as goto jumps!) but maybe need a more deep refactoring to be more nice and readable.

jaescartin1 replied:
Checking the results in 1D, 2D and 3D, both codes generate the same outputs but with a small diferences done by the numerical noise introduced for the conversion from Fortran to C++

jgphpc committed:
Update cmakelist to copy/install the python script

jgphpc commented:
> the `compare_solutions.py` script is not copied to the installation directory

see https://github.com/unibas-dmi-hpc/SPH-EXA/pull/140/commits/c58d9ecadf51dead99474f743fe2cd132f46a835

jaescartin1 committed:
Changed plot outputs from .svg to .png

jgphpc commented:
cscs-ci run

jgphpc commented on src/analyticalSolutions/sedovSolution/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+add_executable(sedovSolution ${SOURCES})
+target_include_directories(sedovSolution PRIVATE ${PROJECT_SOURCE_DIR}/include)
+target_include_directories(sedovSolution PRIVATE ${PROJECT_SOURCE_DIR}/src)
+install(TARGETS sedovSolution RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
```
There is a failure when building `sedovSolution` with clang++:
```
ld.lld: error: undefined symbol: std::filesystem::__cxx11::path::_M_split_cmpts()
>>> referenced by main.cpp
```

This fixes the issue:
```suggestion
target_link_libraries(sedovSolution stdc++fs)
install(TARGETS sedovSolution RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
```

jaescartin1 committed:
Update src/analyticalSolutions/sedovSolution/CMakeLists.txt

Co-authored-by: jgp <jgphpc@users.noreply.github.com>

jaescartin1 committed:
Fixed file naming convention

jaescartin1 committed:
Merge branch 'analytical_sedov' of
https://github.com/jaescartin1/SPH-EXA.git into analytical_sedov

Conflicts:
	src/analyticalSolutions/sedovSolution/CMakeLists.txt

Solved conflict due to a renaming of the file

jaescartin1 committed:
Initialize iteration to -1.
In case the user doesn't specify the iteration it should received an
error

jaescartin1 commented:
Hi @sebkelle1 ,
     I set the iteration value to -1. Like first is check if time>0.0 (default value time=0.0),  if you don't specify time and you give the 'constant.txt' file instead but without iteration you will receive an error.
It make sense to check the iteration=0, if you want to compare the model at time=0. with the analytical solutión
... but I think there is an issue in the 'constants.txt' .  
iter=0 is the initial snapshot, without any movement but the constants file said that the time is t=dt, instead t=0.
Maybe you added t=t+dt before to write the line in the constants file.

jaescartin1 commented:
$ cat constants.txt 
0 1.1e-06 1.1e-06 1 1.00739e-08 1 0 12702448 
1 2.31e-06 1.21e-06 1 3.10683e-08 1 0 12697729 
...

jaescartin1 committed:
Little fix in cmakelist because I changed the name

jaescartin1 committed:
Added --out_path paramter from move the outputs to a tmp folder

jaescartin1 committed:
Added L1 error calculation for rho,u,p,vel,cs

jaescartin1 commented:
All the changes proposed in your comments are in-place.
Additionally I added the L1 error control for the common variables: rho, u, p, vel, cs
Please, review again the changes

jaescartin1 commented:
Here the execution commands for the new version. 
All the output files will be generated in ./bin/
The pictures will be in .png format.
The call also check the rho L1 error

$ git clone --branch "analytical_sedov" https://github.com/jaescartin1/SPH-EXA.git
$ cd SPH-EXA && make -j mpi+omp && ./bin/mpi+omp.app -n 50 -s 200 -w 200 --outDir ./bin/
$ python src/analytical_solutions/compare_solutions.py sedov --binary_file bin/sedov_solution --constants_file ./bin/constants.txt --iteration 200 --nparts 125000 --snapshot_file ./bin/dump_sedov200.txt --out_dir bin/ --error_rho --delta_rho 1.e-0

jgphpc commented:
> I cannot see the python file in the build/ folder. Am I doing something wrong?

no, it will copy it only when running `make install`.
That said, I need to fix an issue in cmake (I'll push).

Btw, `-DBUILD_ANALYTICAL=ON` is on by default.

jgphpc commented:
cscs-ci run

sekelle committed:
fix compare_solutions.py install path

sekelle commented:
cscs-ci run

sekelle commented:
This is ready for merging from my side. As soon as the CI is available again we can merge.

jaescartin1 committed:
Fix L1 errors. Now the values taken are in the right positions

sekelle commented on src/analytical_solutions/sedov_solution/sedov_solution.hpp:
```diff
@@ -41,6 +41,9 @@
 
+#ifndef _SEDOV_SOLUTION_HPP_
+#define _SEDOV_SOLUTION_HPP_
+
```
We don't need these include guards. We have "pragma once" just below

sekelle commented on src/analytical_solutions/sedov_solution/particle.hpp:
```diff
@@ -0,0 +1,62 @@
+ */
+
+#ifndef _SEDOV_SOLUTION_PARTICLE_HPP_
+#define _SEDOV_SOLUTION_PARTICLE_HPP_
```
not needed

sekelle commented on src/analytical_solutions/sedov_solution/io.hpp:
```diff
@@ -29,12 +29,17 @@
  */
 
+#ifndef _SEDOV_SOLUTION_IO_HPP_
+#define _SEDOV_SOLUTION_IO_HPP_
```
not needed

jgphpc commented on src/analytical_solutions/sedov_solution/main.cpp:
```diff
@@ -0,0 +1,221 @@
+ */
+
+#include <string>
+#include <iostream>
```
```suggestion
#include <sstream>
#include <iostream>
```

jgphpc commented on src/analytical_solutions/sedov_solution/io.cpp:
```diff
@@ -0,0 +1,200 @@
+
+#include "io.hpp"
+
+#include <iostream>
```
```suggestion
#include <sstream>
#include <iostream>
```

jaescartin1 committed:
Remove defines, previously added to avoid the inclusion more than one
time

jgphpc commented on src/analytical_solutions/sedov_solution/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+add_executable(sedov_solution ${SOURCES})
+target_include_directories(sedov_solution PRIVATE ${PROJECT_SOURCE_DIR}/include)
+target_include_directories(sedov_solution PRIVATE ${PROJECT_SOURCE_DIR}/src)
+target_link_libraries(sedov_solution stdc++fs)
```
```suggestion
if(CMAKE_OSX_SYSROOT)
    target_link_libraries(sedov_solution stdc++)
else()
     target_link_libraries(sedov_solution stdc++fs)
endif()
```

jaescartin1 committed:
Update src/analytical_solutions/sedov_solution/main.cpp

Co-authored-by: jgp <jgphpc@users.noreply.github.com>

jaescartin1 committed:
Update src/analytical_solutions/sedov_solution/io.cpp

Co-authored-by: jgp <jgphpc@users.noreply.github.com>

jaescartin1 committed:
Update src/analytical_solutions/sedov_solution/CMakeLists.txt

Co-authored-by: jgp <jgphpc@users.noreply.github.com>

jaescartin1 commented:
Ruben is checking the Error L1 in u and cs. Let's wait for the merge till Ruben give the ok

jaescartin1 committed:
cosmetic changes in the error L1 prints

jgphpc committed:
Update compare_solutions.py

jgphpc commented:
⚠️: I pushed compare_solutions.py (reformatting only, no change to the code)

jaescartin1 committed:
Cosmetic changes

jaescartin1 commented:
Perfect, I put my last changes in your reformatted code
